### PR TITLE
AP_Filesystem: allow transfer of mission/fence/rally/param with mavftp

### PR DESCRIPTION
**Pull request by tridge**
This exposes a @MISSION filesystem for upload and download of mission/fence/rally and extends @PARAM filesystem for parameter upload with ftp

mavproxy implementation: https://github.com/ArduPilot/MAVProxy/pull/880

working:
 - [x] mission download
 - [x] mission upload
 - [x] rally download
 - [x] fence download
 - [x] param upload

### CODE REVIEW

**Review by magicrub — COMMENTED** [review]
I breezed through it, looks straight-forward. I look forward to this getting merged!

**Comment by davidbuzz** [comment]
I've had a look through this, like @magicrub , and it looks pretty clean to me.    I'd be more comfortable with all the low-level filesystem functionality being correct if it had auto-tests that exercised all the upload/download and compared the before-upload and the after-download, etc, but that's just me, i know auto-tests aren't really considered 'essential' at this point.

File: libraries/AP_Filesystem/AP_Filesystem_Mission.cpp [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+                if (!r.writebuf->append(nullptr, flen - r.writebuf->get_length())) {
+                    // not enough memory
+                    return -1;
+                }
+            }
+        }
+    }
+    if (r.file_ofs + count > r.writebuf->get_length()) {
+        if (!r.writebuf->append(nullptr, r.file_ofs + count - r.writebuf->get_length())) {
+            return -1;
```
Comment (andyp1per) [inline libraries/AP_Filesystem/AP_Filesystem_Mission.cpp:314]: ENOSPC?

File: libraries/AP_Filesystem/AP_Filesystem_Mission.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    int open(const char *fname, int flags) override;
+    int close(int fd) override;
+    int32_t read(int fd, void *buf, uint32_t count) override;
+    int32_t lseek(int fd, int32_t offset, int whence) override;
+    int32_t write(int fd, const void *buf, uint32_t count) override;
+    int stat(const char *pathname, struct stat *stbuf) override;
+
+private:
+    // only allow up to 4 files at a time
+    static constexpr uint8_t max_open_file = 4;
```
Comment (andyp1per) [inline libraries/AP_Filesystem/AP_Filesystem_Mission.h:35]: FWIW I think constants should be upper case

File: libraries/GCS_MAVLink/MissionItemProtocol_Fence.cpp [hunk 3]
```diff
@@ -50,6 +55,23 @@ MAV_MISSION_RESULT MissionItemProtocol_Fence::get_item(const GCS_MAVLINK &_link,
     ret_packet.y = fenceitem.loc.y;
     ret_packet.z = 0;
 
+    return true;
+}
+
+MAV_MISSION_RESULT MissionItemProtocol_Fence::get_item(const GCS_MAVLINK &_link,
```
Comment (andyp1per) [inline libraries/GCS_MAVLink/MissionItemProtocol_Fence.cpp:61]: const

**Review by andyp1per — APPROVED** [review]
Can't see anything obviously wrong! Two comments:
- Lots of string manipulation, best to run through valgrind if you haven't already
- Its quite a lot of code - conditional compile for 1Mb boards?

File: libraries/AP_Filesystem/AP_Filesystem_Mission.cpp [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+        return MissionItemProtocol_Rally::get_item_as_mission_item(idx, item);
+
+    default:
+        break;
+    }
+    return false;
+}
+
+// get number of items
+uint32_t AP_Filesystem_Mission::get_num_items(enum MAV_MISSION_TYPE mtype) const
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_Filesystem/AP_Filesystem_Mission.cpp:250]: it is const

File: libraries/AP_Filesystem/AP_Filesystem_Mission.h (hunk 2 above)
Comment (tridge) [reply libraries/AP_Filesystem/AP_Filesystem_Mission.h:35]: the other filesystems in same dir use lower case

File: libraries/GCS_MAVLink/MissionItemProtocol_Fence.cpp (hunk 3 above)
Comment (tridge) [reply libraries/GCS_MAVLink/MissionItemProtocol_Fence.cpp:61]: its an inherited API that is non-const

File: libraries/AP_Filesystem/AP_Filesystem_Mission.cpp (hunk 1 above)
Comment (tridge) [reply libraries/AP_Filesystem/AP_Filesystem_Mission.cpp:314]: fixed

File: libraries/AP_Filesystem/AP_Filesystem_Mission.cpp [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+    struct header hdr;
+    if (r.file_ofs == 0 && count >= sizeof(hdr)) {
+        // pre-expand the buffer to the full size when we get the header
+        memcpy(&hdr, buf, sizeof(hdr));
+        if (hdr.num_items < 0xFFFF) {
+            const uint32_t flen = sizeof(hdr) + hdr.num_items * MAVLINK_MSG_ID_MISSION_ITEM_INT_LEN;
+            if (flen > r.writebuf->get_length()) {
+                if (!r.writebuf->append(nullptr, flen - r.writebuf->get_length())) {
+                    // not enough memory
+                    return -1;
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_Filesystem/AP_Filesystem_Mission.cpp:308]: fixed

File: libraries/AP_Filesystem/AP_Filesystem_Mission.cpp [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+        mtype = MAV_MISSION_TYPE_RALLY;
+        return true;
+    }
+    return false;
+}
+
+/*
+  get one item
+ */
+bool AP_Filesystem_Mission::get_item(uint32_t idx, enum MAV_MISSION_TYPE mtype, mavlink_mission_item_int_t &item)
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_Filesystem/AP_Filesystem_Mission.cpp]: done

**Comment by tridge** [comment]
> Its quite a lot of code - conditional compile for 1Mb boards?

what I'd like to do is make the whole mavlink ftp support conditional, and allow disable of it in the custom fw server

**Comment by tridge** [comment]
> Lots of string manipulation, best to run through valgrind if you haven't already

good call btw, there was a socket write of uninitialised memory, fixed now

**Review by andyp1per — APPROVED** [review]

**Pull request merged**